pkg/reconciler/pipelinerun/pipelinerun_test.go

```diff
@@ -4350,6 +4350,188 @@ func TestReconcilePipeline_TaskSpecMetadata(t *testing.T) {
+			Name:      "test-pipeline-run-final-task-results-dag-task-2-xxyyy",
+			Namespace: "foo",
+			OwnerReferences: []metav1.OwnerReference{{
+				Kind:       "PipelineRun",
+				Name:       "test-pipeline-run-final-task-results",
+				APIVersion: "tekton.dev/v1beta1",
+			}},
+			Labels: map[string]string{
+				"tekton.dev/pipeline":     "test-pipeline",
+				"tekton.dev/pipelineRun":  "test-pipeline-run-final-task-results",
+				"tekton.dev/pipelineTask": "dag-task-2",
+			},
+		},
+		Spec: v1beta1.TaskRunSpec{
+			ServiceAccountName: "test-sa",
+			TaskRef:            &v1beta1.TaskRef{Name: "hello-world"},
+		},
+		Status: v1beta1.TaskRunStatus{
+			Status: duckv1beta1.Status{
+				Conditions: duckv1beta1.Conditions{apis.Condition{
+					Type:   apis.ConditionSucceeded,
+					Status: corev1.ConditionFalse,
+				}},
+			},
+		},
+	}}
+
+	d := test.Data{
+		PipelineRuns: prs,
+		Pipelines:    ps,
+		Tasks:        ts,
+		TaskRuns:     trs,
+	}
+	prt := NewPipelineRunTest(d, t)
+	defer prt.Cancel()
+
+	reconciledRun, clients := prt.reconcileRun("foo", "test-pipeline-run-final-task-results", []string{}, false)
+
+	expectedTaskRunName := "test-pipeline-run-final-task-results-final-task-1-9l9zj"
+	expectedTaskRun := v1beta1.TaskRun{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      expectedTaskRunName,
+			Namespace: "foo",
+			OwnerReferences: []metav1.OwnerReference{{
+				Kind:               "PipelineRun",
+				Name:               "test-pipeline-run-final-task-results",
+				APIVersion:         "tekton.dev/v1beta1",
+				Controller:         &trueb,
+				BlockOwnerDeletion: &trueb,
+			}},
+			Labels: map[string]string{
+				"tekton.dev/pipeline":     "test-pipeline",
+				"tekton.dev/pipelineRun":  "test-pipeline-run-final-task-results",
+				"tekton.dev/pipelineTask": "final-task-1",
+			},
+			Annotations: map[string]string{},
+		},
+		Spec: v1beta1.TaskRunSpec{
+			ServiceAccountName: "test-sa-0",
+			TaskRef:            &v1beta1.TaskRef{Name: "final-task"},
+			Params:             []v1beta1.Param{{Name: "finalParam", Value: v1beta1.ArrayOrString{Type: "string", StringVal: "aResultValue"}}},
+			Resources:          &v1beta1.TaskRunResources{},
+			Timeout:            &metav1.Duration{Duration: config.DefaultTimeoutMinutes * time.Minute},
+		},
+	}
+
+	// Check that the expected TaskRun was created
+	actual, err := clients.Pipeline.TektonV1beta1().TaskRuns("foo").List(prt.TestAssets.Ctx, metav1.ListOptions{
+		LabelSelector: "tekton.dev/pipelineTask=final-task-1,tekton.dev/pipelineRun=test-pipeline-run-final-task-results",
+		Limit:         1,
+	})
+
+	if err != nil {
+		t.Fatalf("Failure to list TaskRun's %s", err)
+	}
+	if len(actual.Items) != 1 {
+		t.Fatalf("Expected 1 TaskRuns got %d", len(actual.Items))
+	}
+	actualTaskRun := actual.Items[0]
+	if d := cmp.Diff(expectedTaskRun, actualTaskRun, ignoreResourceVersion); d != "" {
+		t.Errorf("expected to see TaskRun %v created. Diff %s", expectedTaskRunName, diff.PrintWantGot(d))
+	}
+
```

Comment:
continuing on this conversation:

> me: would it make sense to also make sure the overall status of the reconciled run is failure in this case? unless that's something you want to cover in an e2e test instead

> @pritidesai: Reconciled run does not result in failure here, controller logs an error and continues creating taskRun for other final tasks. I have covered the overall pipelineRun status in e2e test.

I think I almost understand, but I think technically the situation is that the controller is waiting for the finally tasks to complete before updating the status to failed, is that right?

> controller logs an error and continues creating taskRun for other final tasks

this reconcile should result in all finally tasks "running" (since they run in parallel) and then the it would the the last reconcile after the finally tasks have finished that would set the overall status to failed, is that right?

Reply:
> I think I almost understand, but I think technically the situation is that the controller is waiting for the finally tasks to complete before updating the status to failed, is that right?

Yes the controller is waiting for the finally tasks to complete before update the status. The status condition need not be necessarily failure. If finally task has a reference to a skipped dag task, the controller sets the status to `completed` after all finally tasks are finished. The controller sets the status to `failed` if any of the finally or a dag task fails. Let me know if this is not clear.